### PR TITLE
Adds gauss weapons

### PR DESCRIPTION
**Pull request by rye-rice**
## About The Pull Request

Adds a powered subtype of ballistic that requires power to shoot. Only adds one weapon at the moment. Almost all of SolGov's armory lore wise is supposed to be this, so expect more in the future

Screenshot of the test gun 
image

Current issues is the sprites suck, but all bullet sprites are kind of bad at the moment anyways.

Ported from [Citadel's](https://github.com/Citadel-Station-13/Citadel-Station-13) Magrifles
Sounds from [TGMC](https://github.com/tgstation/TerraGov-Marine-Corps) 

## Why It's Good For The Game

more unique gun options in the current state of click to win guns

## Changelog
:cl:
add: powered ballistics
fix: fixed a bug where the loading sound effect would play when ejecting a magazine
/:cl:

### CODE REVIEW

**Comment by Fikou** [comment]
that changelog seems fucked up

File: code/modules/projectiles/guns/powered.dm [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	var/overlay_icon_state = "[icon_state]_charge"
+	var/ratio = get_charge_ratio()
+	if(cell)
+		. += "[icon_state]_cell"
+	if(ratio == 0)
+		. += "[icon_state]_cellempty"
+	else
+		if(!shaded_charge)
+			var/mutable_appearance/charge_overlay = mutable_appearance(icon, overlay_icon_state)
+			for(var/i = ratio, i >= 1, i--)
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm]: iirc c++ style loops are slower, why not just for var i in 1 to ratio

File: code/modules/projectiles/guns/powered.dm [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+			to_chat(user, "<span class='notice'>You remove the power cell.</span>")
+			eject_cell(user)
+	return ..()
+
+/obj/item/gun/ballistic/automatic/powered/update_overlays()
+	. = ..()
+	if(!automatic_charge_overlays)
+		return
+	var/overlay_icon_state = "[icon_state]_charge"
+	var/ratio = get_charge_ratio()
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm]: charge_ratio would be a bit more readable

File: code/modules/projectiles/guns/powered.dm [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	if(cell)
+		. += "[icon_state]_cell"
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm:107]: cant find an icon called gauss_cell

File: code/modules/projectiles/guns/powered.dm [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		to_chat(user, "<span class='warning'>\The [C] doesn't seem to fit into \the [src]...</span>")
+		return FALSE
+	if(user.transferItemToLoc(C, src))
+		cell = C
+		to_chat(user, "<span class='notice'>You load the [C] into \the [src].</span>")
+		playsound(src, load_sound, sound_volume, load_sound_vary)
+		update_icon()
+		return TRUE
+	else
+		to_chat(user, "<span class='warning'>You cannot seem to get \the [src] out of your hands!</span>")
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm]: \the is included in [src] im pretty sure

File: code/modules/projectiles/guns/powered.dm [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+/obj/item/gun/ballistic/automatic/powered/proc/insert_cell(mob/user, obj/item/stock_parts/cell/gun/C)
+	if(small_gun && !istype(C, /obj/item/stock_parts/cell/gun/mini))
+		to_chat(user, "<span class='warning'>\The [C] doesn't seem to fit into \the [src]...</span>")
+		return FALSE
+	if(!big_gun && istype(C, /obj/item/stock_parts/cell/gun/large))
+		to_chat(user, "<span class='warning'>\The [C] doesn't seem to fit into \the [src]...</span>")
+		return FALSE
+	if(user.transferItemToLoc(C, src))
+		cell = C
+		to_chat(user, "<span class='notice'>You load the [C] into \the [src].</span>")
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm]: you sure you want "the [C]"?

File: code/modules/projectiles/guns/powered.dm (hunk 4 above)
Comment (Fikou) [reply code/modules/projectiles/guns/powered.dm]: for other lines too

File: code/modules/projectiles/guns/powered.dm [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+/obj/item/gun/ballistic/automatic/powered/examine(mob/user)
+	. = ..()
+	if(cell)
+		. += "<span class='notice'>[src]'s cell is [round(cell.charge / cell.maxcharge, 0.1) * 100]% full.</span>"
+	else
+		. += "<span class='notice'>[src] doesn't seem to have a cell!</span>"
+
+/obj/item/gun/ballistic/automatic/powered/can_shoot()
+	if(QDELETED(cell))
+		return 0
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm]: FALSE?

File: code/modules/projectiles/guns/powered.dm [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	var/cell_type = /obj/item/stock_parts/cell/gun
+	var/charge_sections = 3
+
+	var/shaded_charge = FALSE //if this gun uses a stateful charge bar for more detail
+	var/automatic_charge_overlays = TRUE	//Do we handle overlays with base update_icon()?
+
+	var/internal_cell = FALSE ///if the gun's cell cannot be replaced
+	var/small_gun = FALSE ///if the gun is small and can only fit the small gun cell
+	var/big_gun = FALSE ///if the gun is big and can fit the comically large gun cell
+	var/unscrewing_time = 20 ///Time it takes to unscrew the cell
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm]: SECONDS defines

File: code/modules/projectiles/guns/powered.dm [hunk 8]
```diff
@@ -0,0 +1,123 @@
+/obj/item/gun/ballistic/automatic/powered
+	mag_type = /obj/item/ammo_box/magazine/gauss
+	can_suppress = FALSE
+
+	var/obj/item/stock_parts/cell/cell
+	var/cell_type = /obj/item/stock_parts/cell/gun
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm:6]: you can merge these into one variable maybe? start as a typepath and turn it into an object in init

File: code/game/machinery/recharger.dm [hunk 9]
```diff
@@ -16,7 +16,9 @@
 		/obj/item/gun/energy,
 		/obj/item/melee/baton,
 		/obj/item/ammo_box/magazine/recharge,
-		/obj/item/modular_computer))
+		/obj/item/modular_computer,
+		/obj/item/gun/ballistic/automatic/powered
```
Comment (Fikou) [inline code/game/machinery/recharger.dm]:
```suggestion
		/obj/item/gun/ballistic/automatic/powered,
```

**Review by Fikou — CHANGES_REQUESTED** [review]
i also recommend you change the one letter variable names, but code looks good otherwise

File: code/modules/projectiles/guns/powered.dm (hunk 6 above)
Comment (Fikou) [reply code/modules/projectiles/guns/powered.dm]: for other return 0s in here too

File: code/modules/projectiles/guns/powered.dm (hunk 7 above)
Comment (rye-rice) [reply code/modules/projectiles/guns/powered.dm]: 2 SECONDS

File: code/modules/projectiles/guns/powered.dm (hunk 6 above)
Comment (rye-rice) [reply code/modules/projectiles/guns/powered.dm]: FALSE

File: code/modules/projectiles/guns/powered.dm (hunk 5 above)
Comment (rye-rice) [reply code/modules/projectiles/guns/powered.dm]: fixed

File: code/modules/projectiles/guns/powered.dm (hunk 4 above)
Comment (rye-rice) [reply code/modules/projectiles/guns/powered.dm]: fixed,along with other lines

File: code/modules/projectiles/guns/powered.dm (hunk 3 above)
Comment (rye-rice) [reply code/modules/projectiles/guns/powered.dm:107]: its for future proofing, for when the solgov expansion @PiperDoots is working on is finished

File: code/modules/projectiles/guns/powered.dm (hunk 2 above)
Comment (rye-rice) [reply code/modules/projectiles/guns/powered.dm]: ratioed

File: code/modules/projectiles/guns/powered.dm (hunk 1 above)
Comment (rye-rice) [reply code/modules/projectiles/guns/powered.dm]: changed

File: code/modules/projectiles/guns/powered.dm (hunk 8 above)
Comment (rye-rice) [reply code/modules/projectiles/guns/powered.dm:6]: Doing this did some very strange effects, where it didnt take the charge from the gun cell and fired without it?

File: code/modules/projectiles/guns/powered.dm [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+	if(!shot)
+		return FALSE
+	if(cell.charge < shot.energy_cost * burst_size)
+		return FALSE
+	. = ..()
+
+/obj/item/gun/ballistic/automatic/powered/shoot_live_shot(mob/living/user, pointblank = FALSE, mob/pbtarget, message = 1, stam_cost = 0)
+	var/obj/item/ammo_casing/caseless/gauss/shot = chambered
+	cell.use(shot.energy_cost)
+	. = ..()
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm]: use return ..() at the end of procs

File: code/modules/projectiles/guns/powered.dm [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+/obj/item/gun/ballistic/automatic/powered/can_shoot()
+	if(QDELETED(cell))
+		return FALSE
+
+	var/obj/item/ammo_casing/caseless/gauss/shot = chambered
+	if(!shot)
+		return FALSE
+	if(cell.charge < shot.energy_cost * burst_size)
+		return FALSE
+	. = ..()
```
Comment (Fikou) [inline code/modules/projectiles/guns/powered.dm]: and here

**Review by Fikou — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/powered.dm (hunk 8 above)
Comment (Fikou) [reply code/modules/projectiles/guns/powered.dm:6]: how were you doing it?

Comment (rye-rice) [reply code/modules/projectiles/guns/powered.dm:6]: at first, i was doing it like:

cell = new cell(src)

then,

cell = new(src)

neither of which worked

**Comment by Fikou** [comment]
we can give this a go

**Review by Fikou — APPROVED** [review]

**Pull request merged**